### ISSUE TITLE
Rebalance route finder

### ISSUE THREAD

**Issue by TheBlueMatt** [rust-bitcoin/rust-lightning#959]
We should build a second `get_route`-like method that, instead of seeking to send funds from one node to another, seeks to balance the funds in a set of `ChannelDetails` objects by routing through a `NetworkGraph` back to the same node.

**Comment by abhik-99** [comment]
Hey, if no one is working on this, may I take a shot at it?

**Comment by TheBlueMatt** [comment]
Go for it! I imagine it will be simplest to refactor the existing get_route into a thin wrapper and a helper which does the work and supports both rebalance and regular route finding (or at least is used to calculate a path from our peer back to us without the to-be-balanced channel being used).

**Comment by abhik-99** [comment]
I was actually thinking about implementing a method for [this](https://arxiv.org/pdf/1912.09555.pdf) paper by Rene Pickhardt. But I take it at this time a Just in Time Routing scheme would be favourable? I can follow [c-lightning implementation](https://github.com/lightningd/plugins/pull/66) by cdecker.

The thought of giving the user a choice to select the rebalancing option or finding an alternative route sounds good.

**Comment by TheBlueMatt** [comment]
If we want a parallel route finder (which we could totally do) we should support both routefinders for rebalance-route-finding. For rebalancing we probably don’t want something like the Pickhardt route-finding since we don’t really care about reliability, we just care about minimizing fees.

Pickhardt route-finding we may not want to do anytime soon anyway - it is likely too restrictive in terms of requiring base fees and may overpay in fees, we should first experiment with fee-based scoring of the form “I’m willing to pay X extra in fees to have an N% more reliable payment” which is rather simple to implement. That is something Jeff is currently working on.

**Comment by abhik-99** [comment]
Got it!

**Comment by abhik-99** [comment]
Hey, sorry but I cannot work on this issue at this time. So if anyone is interested, please feel free to take it up.

**Comment by dunxen** [comment]
I can take a look at this if it is  still relevant?

**Comment by TheBlueMatt** [comment]
Yep, go for it! I think there's somewhat less interest in lightning in rebalancing in general but its definitely still something we should support, for users who want it.

**Comment by dunxen** [comment]
Got tied up in Qala.dev stuff, but planning to continue with this in a few days. No ETA though :)

**Comment by dunxen** [comment]
Almost have a PR up for this for early review
